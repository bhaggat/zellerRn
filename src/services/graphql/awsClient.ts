import {Amplify} from 'aws-amplify';
import {generateClient} from 'aws-amplify/api';
import {
  AWS_APPSYNC_GRAPHQL_ENDPOINT,
  AWS_APPSYNC_REGION,
  AWS_APPSYNC_AUTHENTICATION_TYPE,
  AWS_APPSYNC_API_KEY,
} from '@env';

type GraphQLAuthMode =
  | 'apiKey'
  | 'oidc'
  | 'userPool'
  | 'iam'
  | 'identityPool'
  | 'lambda'
  | 'none';

Amplify.configure({
  API: {
    GraphQL: {
      endpoint: AWS_APPSYNC_GRAPHQL_ENDPOINT,
      region: AWS_APPSYNC_REGION,
      apiKey: AWS_APPSYNC_API_KEY,
      defaultAuthMode: AWS_APPSYNC_AUTHENTICATION_TYPE as GraphQLAuthMode,
    },
  },
});

export const awsClient = generateClient();
console.log(awsClient);
