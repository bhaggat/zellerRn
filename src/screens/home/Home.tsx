import React, {useCallback, useEffect, useState} from 'react';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import {Customer, getCustomers} from '../../services/graphql/customers';
import CustomerTypes from '../../components/customer-types/CustomerTypes';
import CustomerCard from '../../components/customer-card/CustomerCard';
import Divider from '../../components/divider/Divider';
import SearchBar from '../../components/searchbar/SearchBar';

export type CustomerType = 'ADMIN' | 'MANAGER';
export const customerTypeOptions: CustomerType[] = ['ADMIN', 'MANAGER'];

export default function Home() {
  const [selectedCustomerType, setSelectedCustomerType] =
    useState<CustomerType>(customerTypeOptions[0]);
  const [customers, setCustomers] = useState<Customer[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [search, setSearch] = useState<string>('');
  const [error, setError] = useState<string | null>(null);

  const keyExtractor = useCallback((customer: Customer) => customer.id, []);

  const getCustomerData = useCallback(async () => {
    setIsLoading(true);
    setError(null);
    try {
      const response = await getCustomers({
        role: selectedCustomerType,
        search,
      });
      if (Array.isArray(response)) {
        setCustomers(response);
      }
    } catch (error) {
      setError('Failed to fetch customer data. Please try again later.');
    } finally {
      setIsLoading(false);
    }
  }, [search, selectedCustomerType]);

  useEffect(() => {
    getCustomerData();
  }, [getCustomerData]);

  return (
    <View style={styles.container}>
      <SearchBar isLoading={isLoading} onSearch={setSearch} />
      <Divider />
      <View style={styles.titleContainer}>
        <Text style={styles.title}>User Types</Text>
      </View>
      <CustomerTypes
        selectedCustomerType={selectedCustomerType}
        onSelectCustomerType={setSelectedCustomerType}
      />
      <Divider />
      <View style={styles.titleContainer}>
        <Text style={styles.title}>{selectedCustomerType} Users</Text>
      </View>
      <View style={styles.listContainer}>
        {isLoading && !customers.length ? (
          <ActivityIndicator size="large" testID="loadingIndicator" />
        ) : error ? (
          <Text style={styles.errorText}>{error}</Text>
        ) : (
          <FlatList
            refreshing={isLoading}
            testID="UsersFlatList"
            onRefresh={getCustomerData}
            keyExtractor={keyExtractor}
            data={customers}
            ItemSeparatorComponent={() => <View style={styles.separator} />}
            ListFooterComponent={<Divider style={styles.footerDivider} />}
            ListEmptyComponent={
              <View style={styles.emptyContainer}>
                <Text style={styles.emptyText}>No users found</Text>
              </View>
            }
            renderItem={({item}) => <CustomerCard customer={item} />}
          />
        )}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    padding: 16,
  },
  title: {
    textTransform: 'capitalize',
    fontSize: 20,
    fontWeight: 'bold',
  },
  titleContainer: {
    marginVertical: 16,
  },
  listContainer: {
    flex: 1,
  },
  separator: {
    height: 10,
  },
  footerDivider: {
    marginTop: 20,
  },
  emptyContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 100,
    width: '100%',
  },
  emptyText: {
    textAlign: 'center',
    color: 'grey',
  },
  errorText: {
    color: 'red',
    textAlign: 'center',
  },
});
