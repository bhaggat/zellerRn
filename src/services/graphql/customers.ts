import {awsClient} from './awsClient';
import {CustomerType} from '../../screens/home/Home';

const listZellerCustomers = `
  query listZellerCustomers($filter: TableZellerCustomerFilterInput, $limit: Int, $nextToken: String) {
    listZellerCustomers(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;

export type Customer = {
  id: string;
  name: string;
  email: string;
  role: string;
};

const LIMIT = 5;

type Options = {
  role: CustomerType;
  search: string;
};

export async function getCustomers(options: Options) {
  try {
    const variables = {
      filter: {
        role: {eq: options.role},
        name: {contains: options.search},
      },
      input: {
        limit: LIMIT,
      },
    };
    const response = await awsClient.graphql({
      query: listZellerCustomers,
      variables,
    });
    if ('data' in response) {
      return response?.data?.listZellerCustomers?.items || [];
    }
    return [];
  } catch (error) {
    return [];
  }
}
