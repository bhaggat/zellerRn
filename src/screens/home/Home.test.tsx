import React, {act} from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import {getCustomers} from '../../services/graphql/customers';
import Home from './Home';

jest.mock('../../services/graphql/customers', () => ({
  getCustomers: jest.fn(),
}));

const mockedGetCustomers =
  require('../../services/graphql/customers').getCustomers;

describe('Home Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('Snapshots', () => {
    test('matches', () => {
      const tree = render(<Home />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Render', () => {
    it('renders the loading indicator while fetching data', async () => {
      mockedGetCustomers.mockResolvedValue([]);
      const {getByTestId} = render(<Home />);
      expect(getByTestId('loadingIndicator')).toBeTruthy();
      await waitFor(() => expect(mockedGetCustomers).toHaveBeenCalledTimes(1));
    });
    it('renders error message if data fetching fails', async () => {
      mockedGetCustomers.mockRejectedValue(new Error('Network error'));
      const {getByText} = render(<Home />);
      await waitFor(() => {
        expect(
          getByText('Failed to fetch customer data. Please try again later.'),
        ).toBeTruthy();
      });
    });
    it('renders customer data when fetched successfully', async () => {
      const mockCustomers = [
        {id: '1', name: 'Dhruv Bhagat'},
        {id: '2', name: 'Kanika Bhagat'},
      ];
      mockedGetCustomers.mockResolvedValue(mockCustomers);
      const {getByText} = render(<Home />);
      await waitFor(() => {
        expect(getByText('Dhruv Bhagat')).toBeTruthy();
        expect(getByText('Kanika Bhagat')).toBeTruthy();
      });
    });
    it('renders empty state when no customers are found', async () => {
      mockedGetCustomers.mockResolvedValue([]);
      const {getByText} = render(<Home />);
      await waitFor(() => {
        expect(getByText('No users found')).toBeTruthy();
      });
    });
  });

  describe('Interactions', () => {
    it('updates the customer list when the search term changes', async () => {
      const mockCustomers = [{id: '1', name: 'Dhruv Bhagat'}];

      mockedGetCustomers
        .mockResolvedValueOnce(mockCustomers)
        .mockResolvedValueOnce([]);

      const {getByPlaceholderText, getByText} = render(<Home />);
      await waitFor(() => {
        expect(getByText('Dhruv Bhagat')).toBeTruthy();
      });
      const searchBar = getByPlaceholderText('Search...');
      fireEvent.changeText(searchBar, 'Test');
      fireEvent.changeText(searchBar, 'Test1');
      await act(async () => {
        jest.advanceTimersByTime(750);
      });
      await waitFor(() => {
        expect(getByText('No users found')).toBeTruthy();
      });
    });

    it('changes customer type and fetches new data', async () => {
      const adminCustomers = [{id: '1', name: 'Admin User'}];
      const managerCustomers = [{id: '2', name: 'Manager User'}];

      mockedGetCustomers
        .mockResolvedValueOnce(adminCustomers)
        .mockResolvedValueOnce(managerCustomers);

      const {getByText} = render(<Home />);

      await waitFor(() => {
        expect(getByText('Admin User')).toBeTruthy();
      });

      const managerTab = getByText('MANAGER');
      fireEvent.press(managerTab);

      await waitFor(() => {
        expect(getByText('Manager User')).toBeTruthy();
      });
    });

    it('refreshes customer list on pull-to-refresh', async () => {
      const mockCustomers = [{id: '1', name: 'Dhruv Bhagat'}];

      mockedGetCustomers.mockResolvedValue(mockCustomers);
      const {getByText, getByTestId} = render(<Home />);

      await waitFor(() => {
        expect(getByText('Dhruv Bhagat')).toBeTruthy();
      });

      const flatList = getByTestId('UsersFlatList');
      fireEvent(flatList, 'refresh');

      await waitFor(() => {
        expect(mockedGetCustomers).toHaveBeenCalledTimes(2);
      });
    });
  });
});
