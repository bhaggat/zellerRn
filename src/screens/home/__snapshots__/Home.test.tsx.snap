// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Snapshots matches 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "paddingHorizontal": 8,
        "paddingVertical": 4,
        "width": "100%",
      }
    }
  >
    <TextInput
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search..."
      placeholderTextColor="black"
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "height": 40,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#E2E8F0",
          "height": 1,
          "marginVertical": 10,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "textTransform": "capitalize",
        }
      }
    >
      User Types
    </Text>
  </View>
  <View
    style={
      {
        "gap": 4,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E8F2FB",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 10,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#E2E8F0",
              "borderWidth": 1,
              "justifyContent": "center",
            },
            {
              "borderRadius": 10,
              "height": 20,
              "width": 20,
            },
            {
              "borderColor": "#0370CE",
            },
            undefined,
          ]
        }
        testID="radioButton"
      >
        <View
          style={
            {
              "backgroundColor": "#0370CE",
              "borderRadius": 5,
              "height": 10,
              "width": 10,
            }
          }
          testID="innerRadioButton"
        />
      </View>
      <Text
        style={
          {
            "fontSize": 16,
            "textTransform": "capitalize",
          }
        }
      >
        ADMIN
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 10,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#E2E8F0",
              "borderWidth": 1,
              "justifyContent": "center",
            },
            {
              "borderRadius": 10,
              "height": 20,
              "width": 20,
            },
            false,
            undefined,
          ]
        }
        testID="radioButton"
      />
      <Text
        style={
          {
            "fontSize": 16,
            "textTransform": "capitalize",
          }
        }
      >
        MANAGER
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#E2E8F0",
          "height": 1,
          "marginVertical": 10,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "textTransform": "capitalize",
        }
      }
    >
      ADMIN
       Users
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <ActivityIndicator
      size="large"
      testID="loadingIndicator"
    />
  </View>
</View>
`;

exports[`Home Component Snapshots matches the snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "paddingHorizontal": 8,
        "paddingVertical": 4,
        "width": "100%",
      }
    }
  >
    <TextInput
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search..."
      placeholderTextColor="black"
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "height": 40,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#E2E8F0",
          "height": 1,
          "marginVertical": 10,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "textTransform": "capitalize",
        }
      }
    >
      User Types
    </Text>
  </View>
  <View
    style={
      {
        "gap": 4,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E8F2FB",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 10,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#E2E8F0",
              "borderWidth": 1,
              "justifyContent": "center",
            },
            {
              "borderRadius": 10,
              "height": 20,
              "width": 20,
            },
            {
              "borderColor": "#0370CE",
            },
            undefined,
          ]
        }
        testID="radioButton"
      >
        <View
          style={
            {
              "backgroundColor": "#0370CE",
              "borderRadius": 5,
              "height": 10,
              "width": 10,
            }
          }
          testID="innerRadioButton"
        />
      </View>
      <Text
        style={
          {
            "fontSize": 16,
            "textTransform": "capitalize",
          }
        }
      >
        ADMIN
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 10,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#E2E8F0",
              "borderWidth": 1,
              "justifyContent": "center",
            },
            {
              "borderRadius": 10,
              "height": 20,
              "width": 20,
            },
            false,
            undefined,
          ]
        }
        testID="radioButton"
      />
      <Text
        style={
          {
            "fontSize": 16,
            "textTransform": "capitalize",
          }
        }
      >
        MANAGER
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#E2E8F0",
          "height": 1,
          "marginVertical": 10,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "textTransform": "capitalize",
        }
      }
    >
      ADMIN
       Users
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <ActivityIndicator
      size="large"
      testID="loadingIndicator"
    />
  </View>
</View>
`;
